packages/console-extension/src/index.ts

```diff
@@ -405,6 +411,26 @@ async function activateConsole(
       void tracker.save(panel);
     });
 
+    if (options.subshell) {
+      panel.sessionContext.ready
+        .then(async () => {
+          // Ensure kernel has received kernel_info.
+          await panel.sessionContext.session!.kernel!.info;
+          const future =
+            panel.sessionContext.session!.kernel!.requestCreateSubshell({});
```

Comment:
> I suppose the subshells would only ever work on the initial session and initial kernel and not on subsequent ones?

Yes. If the session or kernel change and they are valid then the console will still be a functioning console, but not to the initial subshell.

I have not considered what to do outside of that. If the kernel changes so it is no longer a subshell it could be argued that the console should no longer exist, but closing seems like a very bad thing to do to a user.

Reply:
> but closing seems like a very bad thing to do to a user.

I agree.

> If the kernel changes so it is no longer a subshell

But should it? If user changes kernel, should the console become a subshell console on the new kernel?

Like, many workflows involve "Restart and run all" or "Restart and run up to selected cell". Do we want to require user to use `New Subshell Console for Notebook` command each time after restarting?

Reply:
I think both options are reasonable and it is not obvious to me which is more appropriate. We are in a catch 22 situation with this subshell console. We need to make subshells available in some way to the small number of interested users so they can try it out and discover how they want to interact with them, but we need that information now before we know what it is.

This is different from the remainder of this PR which is just exposing subshells in `KernelConnection`. That enables us to use them for some concrete, easy to understand use cases such as implementing `ipywidgets` and the debugger using subshells if available, eliminating a number of problems caused by blocking without the end user needing to even know that subshells are being used.

Maybe I have approached this PR incorrectly, trying to combine both when they are separate units of functionality. The core subshells in `KernelConnection` doesn't need to be behind a `--kernel-subshells` flag, it should always be present. The subshell console could be separate, either a separate PR behind its own flag, or maybe even in a separate frontend extension although that will invariably require more maintenance in the long run.

Reply:
No, I think this PR is fine. I was just a suggesting a tiny modification, wrapping these lines with:

```ts
panel.sessionContext.kernelChanged.connect(() => {
   // do the registration of subshell with future 
})
```

> The core subshells in `KernelConnection` doesn't need to be behind a `--kernel-subshells` flag, it should always be present

I agree.

> The subshell console could be separate, either a separate PR behind its own flag

No, I think it is good to have it here as it gives a concrete path to testing this and allows for the integration test.

In any case, while I think this is what users would prefer, if it is not clear for you I call it a draw and let's move forward merging this PR once all the other bits are resolved :)